libs/api-catalogue/services/src/lib/provider.service.ts

```diff
@@ -0,0 +1,68 @@
+import { Provider } from '@island.is/api-catalogue/types'
+import {
+  MetaservicesApi,
+  XroadIdentifier,
+  XroadIdentifierIdObjectTypeEnum,
+} from '../../gen/fetch/xrd'
+import { ProviderType } from '@island.is/api-catalogue/consts'
+import { logger } from '@island.is/logging'
+import { exceptionHandler } from './utils'
+import { Injectable } from '@nestjs/common'
+
+@Injectable()
+export class ProviderService {
+  constructor(private readonly xrdMetaService: MetaservicesApi) {}
+
+  async getPrivateProviders(): Promise<Array<Provider>> {
+    return this.getProviders(ProviderType.PRIVATE)
+  }
+
+  async getProtectedProviders(): Promise<Array<Provider>> {
+    return this.getProviders(ProviderType.PROTECTED)
+  }
+
+  async getPublicProviders(): Promise<Array<Provider>> {
+    return this.getProviders(ProviderType.PUBLIC)
+  }
+
+  private async getProviders(
+    providerType: ProviderType,
+  ): Promise<Array<Provider>> {
+    let providers: Array<Provider> = []
+
+    logger.info(`Finding ${providerType} providers`)
+
+    try {
+      const filter = new RegExp(`.*${providerType}$`)
+      const xrdClients = await this.xrdMetaService.listClients({})
+
+      if (xrdClients && xrdClients.member && xrdClients.member.length > 0) {
+        providers = xrdClients.member
+          .filter((item: XroadIdentifier): boolean => {
+            return (
+              item.id?.objectType ===
+                XroadIdentifierIdObjectTypeEnum.SUBSYSTEM &&
+              filter.test(item.id?.subsystemCode?.toLowerCase() ?? '')
+            )
+          })
+          .map(
+            (item: XroadIdentifier): Provider => {
+              return {
+                type: providerType,
+                xroadInstance: item.id?.xroadInstance ?? '',
+                memberClass: item.id?.memberClass ?? '',
+                memberCode: item.id?.memberCode ?? '',
+                subsystemCode: item.id?.subsystemCode ?? '',
+              }
+            },
+          )
+      }
+    } catch (err) {
+      exceptionHandler(err)
```

Comment:
I am concerned with this pattern of swallowing exceptions and returning successfully. Aren't these critical exceptions? Shouldn't they be propagated?

Reply:
This is a valid point. But the XRoad api will throw exceptions if there is some connection issues with other Xroad clients or if rest service are registered manually instead of using openapi. We can solve these issues if we add usage of the new Admin API from XRoad and that is on the roadmap for these services. But until we can do that the idea was to catch these errors, log them but return empty answers, so the whole process won't crash because of one client or one service being misconfigured. 

But I can change the exceptionHandler function to throw the error if it is not XRoad communication error, so other unexpected errors won't be swallowed?